### PR TITLE
Add `--show-diff` cli option

### PR DESCRIPTION
**Pull request by skarab42**
This PR add the `--show-diff` cli option. 

image

Some limitations:
- The `typeToString` method of the `TypeChecker` does not parse all types in the same way. For example, an interface will return its name and not its contents, whereas an type alias will return its contents, which makes it difficult to compare the two types by string comparison.
- The `typeToString` method of the `TypeChecker` does not allow the indentation of the inspected/resolved type to be preserved, so at the moment the types are displayed inline which is not very readable on complex types.

TODO:
- Implement a normalized `typeToString` function.

### CODE REVIEW

File: source/cli.ts
Comment (tommy-mitchell) [inline source/cli.ts:51]: If we're updating the CLI, may as well consolidate these variables:

```suggestion
		const {typingsFile, testFiles, showDiff} = cli.flags;
```

Comment (tommy-mitchell) [reply source/cli.ts:51]: Could also add a description in the help text:

https://github.com/SamVerschueren/tsd/blob/09c0d17365c7585dd8d77f4c0d454d63d0c58e72/source/cli.ts#L16-L18

```
--show-diff    Show type error diffs         [Default: don't show]
```

File: source/test/diff.ts
Comment (tommy-mitchell) [inline source/test/diff.ts:33]:
```suggestion
	const {exitCode, stderr} = await t.throwsAsync<ExecaError>(execa('dist/cli.js', [file, '--show-diff']));
```

Comment (skarab42) [reply source/test/diff.ts:33]: `meow` accepts both syntaxes, but if you prefer this I'm fine with it ;)

**Comment by skarab42** [comment]
Currently the feature is only implemented in `expectType` and needs to be implemented in the other helpers that support it. WIP

**Comment by skarab42** [comment]
> TODO:
> 
> * Implement a normalized `typeToString` function.

I can't work on this feature any time soon... Anyway I was thinking of doing it in another PR. So I won't add anything more in this PR. Waiting for your feedback for eventual changes.

File: source/cli.ts
Comment (tommy-mitchell) [inline source/cli.ts:19]:
```suggestion
	  --show-diff      Show type error diffs         [Default: don't show]
```

File: source/lib/utils/make-diagnostic-with-diff.ts
Comment (tommy-mitchell) [inline source/lib/utils/make-diagnostic-with-diff.ts:42]: Did you mean to add documentation here?

**Review by tommy-mitchell — APPROVED** [review]

File: source/lib/utils/make-diagnostic-with-diff.ts
Comment (skarab42) [reply source/lib/utils/make-diagnostic-with-diff.ts]: Yep ;)

File: source/lib/formatter.ts
Comment (tommy-mitchell) [inline source/lib/formatter.ts:19]:
```suggestion
 * @param diagnostics - List of TypeScript diagnostics.
 * @param showDiff - Display difference between expected and received types.
```

File: source/lib/formatter.ts
Comment (tommy-mitchell) [inline source/lib/formatter.ts:40]: Nitpicky, but I think adding the diagnostic should go after creating the diff, since L52 modifies `diagnostic.message`. (I know that `Map`s give you a mutable reference, but imo it's clearer that the mutation occurs before the `diagnostic` is added.)

File: source/lib/formatter.ts
Comment (tommy-mitchell) [inline source/lib/formatter.ts:45]: Could the replacement go in `makeDiagnosticWithDiff()`?

File: source/cli.ts
Comment (tommy-mitchell) [inline source/cli.ts:52]: Coming back to this, realized I missed that you can rename during destructing:

```suggestion
		const {typings: typingsFile, files: testFiles, showDiff} = cli.flags;
```

**Review by tommy-mitchell — APPROVED** [review]
Sorry for taking so long to review. Besides some minor things, this looks good to me. How were you thinking of doing a normalized `typeToString` function? I could try implementing it.

File: source/lib/formatter.ts
Comment (skarab42) [reply source/lib/formatter.ts]: I removed the `UseSingleQuotesForStringLiteralType` flag in `typeToString` it's better because it keeps the default behavior of tsc so no need to type single quotes.

**Comment by skarab42** [comment]
> How were you thinking of doing a normalized `typeToString` function? I could try implementing it.

You have to explore the AST of the type if it is not stringifiable, you have to browse its properties and format them. It's a huge job with a lot of edge cases to take into account.

**Comment by skarab42** [comment]
@tommy-mitchell I don't know why the CI doesn't pass... I don't see any difference in the output and all the tests run locally under Windows and WLS. Sorry but I can't investigate further.

File: source/lib/utils/make-diagnostic-with-diff.ts
Comment (tommy-mitchell) [inline source/lib/utils/make-diagnostic-with-diff.ts:35]: Maybe the comments should be removed before merging?

**Review by tommy-mitchell — APPROVED** [review]

**Comment by tommy-mitchell** [comment]
Seems like the CI is failing because the `diff cli` test reports the location of the errors, but doesn't locally:

```diff
+'dist/test/fixtures/diff/index.test-d.ts:8:0',
 '✖   8:0  Parameter type { life?: number | undefined; } is declared too wide for argument type { life: number; }.',
 '',
 '- { life?: number | undefined; }',
 '+ { life: number; }',
 // ...
```

File: source/test/diff.ts
Comment (tommy-mitchell) [inline source/test/diff.ts:42]: Which is what this line is supposed to remove, right?

Maybe add this snippet to see what the CI reports is the first line:

```ts
console.log(stderr.trim().split('\n').at(0));
```

(Or log the whole array if that's easier for debugging)

Comment (skarab42) [reply source/test/diff.ts]: Thanks for the hint ;) Fixed, this is because of the addition of the error message and stack trace recently added in https://github.com/SamVerschueren/tsd/blob/621aad986c636ff511293706455720b006598945/source/cli.ts#L61 which adds some lines to the output.

**Review by tommy-mitchell — APPROVED** [review]

**Comment by tommy-mitchell** [comment]
Looks good to me.

**Pull request merged**
